### DISCUSSION TITLE
How to properly read a Json string that may be null in some cases?

### DISCUSSION THREAD

**Discussion by TyloowNv** [nlohmann/json#3139] [Q&A] [answered]
Hello,

I have some json format texts received as server response and would like to use this tool to parse them. The issue is if the value in a string field is optional and not provided from the server side, client side will receive a null rathe than an empty string. This currently throws exceptions while parsing.

For example this is the code where we read the title of a layer.

`bh.layers[j].name = js.at(i)["th_layers"].at(j)["layer_title"].get<string>().c_str();`

However, in the json format, sometime the server response could be this. And this can be parsed without any issue.
`"layer_title": "Fill Material",`
But sometimes it becomes this and throws an exception.
`"layer_title": null,`

Is there a better way of parsing this null and give a default value(like "" for string and 0 for int) rather than using try and catch?

Thanks a lot!

**Comment by gregmarr** [comment, marked as answer]
`js.at(i)["th_layers"].at(j)["layer_title"]`

This is what creates the `null` `"layer_title"`.  You need to check if it exists before you try to get its value.

If `js` is not `const` then accessing a non-existent object value creates it, while if it is `const` it's undefined behavior and asserts.

https://json.nlohmann.me/features/element_access/default_value/

**Reply by TyloowNv** [reply]
Thanks for your reply! I didn't realize you replied so fast!

I tried to do something like this in case the response in layer_title is null:

`auto str = js.at(i)["th_layers"].at(j)["layer_title"]`
`std::string str = js.at(i)["th_layers"].at(j).value("layer_title", "");

js is not const in my code.
The 1st method works, but I'll need to check if (str.IsNull()) to deal with it.
The 2nd method was supposed to give it a default empty string if gets null. But it triggers a 302 error, "type must be string, but is " + std::string(j.type_name()), j))

How should I read it properly with a certain type (int, double or string) when the value could be a null?

**Reply by nlohmann** [reply]
The error occurs, because the type of the JSON value needs to fit the default value.

```cpp
json j;
j["foo"] = 1;

// this does not work, because the value at "foo" is a number, but the default value a string
j.value("foo", "default value");
```

If you can't guarantee the value, you need to write your own function. `is_null` etc. can help here.

**Reply by gregmarr** [reply]
So I assumed that the `null` was being created by your non-const access of a non-existent value.  In that case, the `.value()` would work.  However, if the `null` is actually coming from the server, then you will get the type mismatch there.  In that case, you will need to check the existence and type of `"layer_title"`.  Something like this (not tested, or even compiled):

    auto const &obj = js.at(i)["th_layers"].at(j);
    auto const &layer_title = obj.find("layer_title");
    if  ((layer_title != obj.end()) && layer_title.is_string()) {
        bh.layers[j].name = layer_title.get<string>().c_str();
    }

**Reply by TyloowNv** [reply]
Thanks! I found it easier just to create a new "value" function but always return an empty string if received null.

**Comment by tangtang666** [comment]
I think if use get value by type,it alse get the type default instead of null.
```
//in this case
int code = recv_msg["test"].get<int>()
```
